lib/ansible/cli/doc.py

```diff
@@ -70,76 +73,90 @@ def run(self):
 
         super(DocCLI, self).run()
 
+        plugin_type = self.options.type
+
+        # choose plugin type
+        if plugin_type == 'cache':
+            loader = cache_loader
+        elif plugin_type == 'callback':
+            loader = callback_loader
+        elif plugin_type == 'connection':
+            loader = connection_loader
+        elif plugin_type == 'lookup':
+            loader = lookup_loader
+        elif plugin_type == 'strategy':
+            loader = strategy_loader
+        else:
+            loader = module_loader
+
+        # add to plugin path from command line
         if self.options.module_path is not None:
             for i in self.options.module_path.split(os.pathsep):
-                module_loader.add_directory(i)
+                loader.add_directory(i)
 
         # list modules
         if self.options.list_dir:
-            paths = module_loader._get_paths()
+            paths = loader._get_paths()
             for path in paths:
-                self.find_modules(path)
+                self.find_plugins(path)
 
-            self.pager(self.get_module_list_text())
+            self.pager(self.get_plugin_list_text(loader))
             return 0
 
         # process all modules
-        if self.options.all_modules:
-            paths = module_loader._get_paths()
+        if self.options.all_plugins:
+            paths = loader._get_paths()
             for path in paths:
-                self.find_modules(path)
-            self.args = sorted(set(self.module_list) - module_docs.BLACKLIST_MODULES)
+                self.find_plugins(path)
+            self.args = sorted(set(self.plugin_list))
+
+            bkey = plugin_type.upper()
+            if bkey in plugin_docs.BLACKLIST and plugin in plugin_docs.BLACKLIST[bkey]:
+                self.args = self.args - plugin_docs.BLACKLIST[bkey]
```

Comment:
Probably what you want to do is move the call to sorted(self.args) from line 111 to line 116.